### PR TITLE
Change Tactics To Return Actions Instead Of Intents

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Changed Tactics to return `Action`s instead of `Intent`s

### Testing Done

Unit tests, tested running `full_system` with grsim and running `example_play`.

### Resolved Issues
Part of #773 

### Length Justification
This is about as short as I could get it. Unfortunately once you change the signature in `Tactic.h`, there's not a whole lot you can do to make it shorter.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/action/action.cpp [hunk 1]
```diff
@@ -43,6 +43,11 @@ std::unique_ptr<Intent> Action::getNextIntent()
     return next_intent;
 }
 
+std::optional<Robot> Action::getRobot()
+{
+    return robot;
+}
+
 void Action::calculateNextIntentWrapper(IntentCoroutine::push_type &yield)
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/action.cpp:51]: Can we not delete this and the IntentCoroutine typedef?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good, just couple comments

File: src/software/ai/hl/stp/stp.cpp [hunk 2]
```diff
@@ -21,7 +21,7 @@ STP::STP(std::function<std::unique_ptr<Play>()> default_play_constructor,
 {
 }
 
-std::vector<std::unique_ptr<Intent>> STP::getIntents(const World& world)
+void STP::updateCurrentPlay(const World& world)
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/hl/stp/stp.cpp:24]: Yeah, I was just trying to understand what was going on here, and one way I do that is by refactoring until it makes sense.

File: src/software/ai/hl/stp/action/action.cpp (hunk 1 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/action/action.cpp:51]: Actions still run coroutines, it's just that we get the intents from them in stp now, rather then in tactics. So we still need both.

**Comment by MathewMacDougall** [comment]
@akhilveeraghanta @EvanMorcom @jsn0716 please review

File: src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.cpp [hunk 3]
```diff
@@ -31,14 +31,11 @@ double MoveTestTactic::calculateRobotCost(const Robot &robot, const World &world
...
-void MoveTestTactic::calculateNextIntent(IntentCoroutine::push_type &yield)
+void MoveTestTactic::calculateNextAction(ActionCoroutine::push_type &yield)
 {
     do
     {
-        yield(std::make_unique<MoveIntent>(this->robot->id(), this->destination,
-                                           Angle::zero(), 0.0, 0, DribblerEnable::OFF,
-                                           MoveType::NORMAL, AutokickType::NONE,
-                                           BallCollisionType::AVOID));
+        yield(std::make_shared<MoveAction>());
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.cpp:38]: This should still return a proper MoveIntent?

File: src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.cpp [hunk 4]
```diff
@@ -15,11 +15,11 @@ double StopTestTactic::calculateRobotCost(const Robot &robot, const World &world
     return 0.5;
 }
 
-void StopTestTactic::calculateNextIntent(IntentCoroutine::push_type &yield)
+void StopTestTactic::calculateNextAction(ActionCoroutine::push_type &yield)
 {
     do
     {
-        yield(std::make_unique<StopIntent>(this->robot->id(), false, 0));
+        yield(std::make_shared<StopAction>());
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.cpp:22]: This should yield a proper StopIntent?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/ai/hl/stp/action/kick_action.h [hunk 5]
```diff
@@ -49,6 +49,27 @@ class KickAction : public Action
     void updateControlParams(const Robot &robot, Point kick_origin, Angle kick_direction,
                              double kick_speed_meters_per_second);
 
+    /**
+     * Get the direction this action wants to kick in, relative to the robot kicking
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/hl/stp/action/kick_action.h]: mmk, changed as suggested.

File: src/software/ai/hl/stp/action/movespin_action.cpp [hunk 6]
```diff
@@ -17,6 +17,11 @@ void MoveSpinAction::updateControlParams(const Robot& robot, Point destination,
     this->final_linear_speed = final_linear_speed;
 }
 
+Point MoveSpinAction::getDestination()
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/hl/stp/action/movespin_action.cpp:20]: I'd prefer to keep the scope of this PR to a minimum, I only added the functions required to achieve equivalency with the existing code.

File: src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.cpp (hunk 3 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.cpp:38]: - this is a tactic, so it yields an action now?
- what do you mean by "proper"? This is intended to be a test tactic, so it's implementation is informed by the test requirements, which just require this. I think adding anything extra without have specific reasons for doing some from the test side is more confusing if anything.

File: src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.cpp (hunk 4 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.cpp:22]: Similar comment here to the one in `move_test_tactic.cpp`.

File: src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.cpp:38]: Oh sorry I totally misread this, thought it was returning an empty / null shared ptr. No issues here.

**Review by MathewMacDougall — APPROVED** [review]
LGTM

File: src/software/ai/hl/stp/action/action.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/action/action.cpp:51]: So we're using another layer of coroutines to make tactics return actions?

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**